packages/core/src/indexing-store/historical.ts

```diff
@@ -753,133 +793,169 @@ export const createHistoricalIndexingStore = ({
     ),
     async flush() {
       await queue.add(async () => {
-        const promises: Promise<any>[] = [];
+        let cacheSize = 0;
+        for (const c of cache.values()) cacheSize += c.size;
 
-        for (const [table, tableCache] of cache) {
-          const entries = Array.from(tableCache.values());
+        const flushIndex =
+          totalCacheOps -
+          cacheSize * (1 - common.options.indexingCacheFlushRatio);
+        const shouldDelete = cacheBytes > maxBytes;
+        if (shouldDelete) isDatabaseEmpty = false;
 
-          const batchSize = Math.round(
-            common.options.databaseMaxQueryParameters /
-              Object.keys(getTableColumns(table)).length,
-          );
+        const promises: Promise<void>[] = [];
 
-          const insertValues = entries
+        for (const [table, tableCache] of cache) {
+          const removableInsertValues = Array.from(tableCache.values())
+            .filter(
+              (e) =>
+                e.type === EntryType.INSERT &&
+                shouldDelete &&
+                e.operationIndex < flushIndex,
+            )
+            .map((e) => e.row) as InsertEntry["row"][];
+
+          const removableUpdateValues = Array.from(tableCache.values())
+            .filter(
+              (e) =>
+                e.type === EntryType.UPDATE &&
+                shouldDelete &&
+                e.operationIndex < flushIndex,
+            )
+            .map((e) => e.row) as UpdateEntry["row"][];
+
+          // remove outdated entries, making them eligible for garbage collection
+          for (const [key, entry] of tableCache) {
+            if (shouldDelete && entry.operationIndex < flushIndex) {
+              tableCache.delete(key);
+              cacheBytes -= entry.bytes;
+            }
+          }
+
+          const nonRemovableInsertValues = Array.from(tableCache.values())
             .filter((e) => e.type === EntryType.INSERT)
-            .map((e) => e.row);
+            .map((e) => e.row) as UpdateEntry["row"][];
 
-          const updateValues = entries
+          const nonRemovableUpdateValues = Array.from(tableCache.values())
             .filter((e) => e.type === EntryType.UPDATE)
-            .map((e) => e.row);
+            .map((e) => e.row) as InsertEntry["row"][];
 
-          if (insertValues.length > 0) {
+          for (const [, entry] of tableCache) {
+            entry.type = EntryType.FIND;
+          }
+
+          const insertValues = removableInsertValues.concat(
+            nonRemovableInsertValues,
+          );
+          const updateValues = removableUpdateValues.concat(
+            nonRemovableUpdateValues,
+          );
+
+          const insertSize = insertValues.length;
+          const updateSize = updateValues.length;
+          const insertStream = getReadableStream(table, insertValues);
+          const updateStream = getReadableStream(table, updateValues);
```

Comment:
Is there another opportunity to reduce peak mem usage here by handling the inserts first, then the updates? Again, just an idea - hard to tell what's big/small and what's GC'able here.

I guess the concurrency would get slightly worse.

Reply:
It might, but the case we are really optimizing for is the first call to `flush()`, where all entries are `insert` and none are `find`. 

In this case all entries are serialized into csv, and only 35% of the original entries are garbage collectable. In subsequent call, a good portion of entries will be "find", which don't have to get serialized/copied.